### ISSUE TITLE
Memory requirements

### ISSUE THREAD

**Issue by DagothRa** [KoboldAI/KoboldAI-Client#16]
Something I've noticed is that the memory requirements for the same AI model seem higher for KoboldAI than for CloverEdition. My system has 16 GB system memory, and 8 GB onboard video memory (with an additional 8 GB shared memory available).

So, for example, using GPT-Neo 2.7b.
System memory usage for Clover Edition loading GPT-Neo 2.7b climbs gradually up to 11.5 GB memory, before dropping back down to about 1.2 GB. Meanwhile, at the same time, GPU memory usage goes up to an additional 6.5 GB used, before finally dropping down to about 5.5 GB overhead once it finishes loading.
When trying to load GPT-Neo 2.7b in KoboldAI, the system memory usage climbs up fairly rapidly to over 12 GB, while the GPU memory doesn't budge. My computer then hangs, going almost completely unresponsive, even the clock not updating at all, though every once in a while (every 30 seconds or so, maybe?) the mouse cursor will move a tick if I am moving the mouse. I end up being unable to even kill the process due to the unresponsiveness and have to power cycle my computer. Presumably this is due to my system memory just getting overloaded.

When I load GPT-2 in KoboldAI, as per my previous Issue, it does noticeably start loading into GPU memory almost immediately, so I'm not sure why that seems to be different from GPT-Neo 2.7b loading in KoboldAI.

From the text on the Clover Edition readme page, they mention using 16-bit instead of 32-bit; is that maybe something that would help in KoboldAI?

I'll add that I really like the browser interface and additional features of KoboldAI, and really hate the commandline interface of Clover Edition. I'm just hoping that I can end up running GPT-Neo 2.7b on KoboldAI like I can in Clover Edition.

**Comment by KoboldAI** [comment]
Check your installed version of PyTorch. Models in KoboldAI are initialized through transformers pipelines; while the GPT-2 variants will load and run without PyTorch installed at all, GPT-Neo requires torch even through transformers. As for total memory usage, I'd recommend installing finetune's transformers branch, it'll reduce VRAM usage for Neo-2.7B from 12GB down to about 6.5-7.

pip install https://github.com/finetuneanon/transformers/archive/refs/heads/gpt-neo-dungeon-localattention1.zip

**Comment by finetuneanon** [comment]
Running the models in half precision should work and make neo 2.7B be able to run in 8GB with barely any impact on accuracy. My transformers fork should already be casting the model to half precision automatically as well, unless it's getting cast back to fp32 after loading somehow.

**Comment by DagothRa** [comment]
> pip install https://github.com/finetuneanon/transformers/archive/refs/heads/gpt-neo-dungeon-localattention1.zip

Interesting. I ran that, and then tried again. This time, with GPT-Neo 2.7b, it immediately started loading it into GPU memory, and in the end successfully loaded on my system without causing it to hang, almost identical to CloverEdition loading GPT-Neo 2.7b. Awesome!
Just one issue, though: Where CloverEdition's system memory usage dropped back down to 1.2 GB once it finished loading the AI model, KoboldAI's system memory usage is staying up at nearly 11 GB still allocated to it. Between it and Windows itself, that's not leaving much available for other stuff.

**Comment by DagothRa** [comment]
I definitely think that should be added to the project readme, it makes the larger models playable on a much wider range of hardware with the greatly reduced memory requirements.

**Comment by henk717** [comment]
0.16 has been released with breakmodel allowing you to further manage memory, the readme has also been overhauled so people are pointed towards a version most suitable for them which is Finetuneanon's (newer) version for GPU users.
